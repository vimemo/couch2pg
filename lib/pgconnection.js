import purl from 'url'
import knex from 'knex'

const ensureDatabaseExists = async (url) => {
  const opts = purl.parse(url)
  const dbname = opts.path.slice(1)
  const conn = knex(url.replace(opts.path, ''))
  try {
    await conn.raw(`CREATE DATABASE \"${dbname}\"`)
  } catch(err){} //db already exists - ignore
  conn.destroy()
}

export default (url) => {
  ensureDatabaseExists(url)
  return knex({client: 'pg', connection: url})
}
