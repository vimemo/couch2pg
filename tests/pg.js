import migrate from '../lib/pgmigration'
import Pouch from '../lib/Pouch'
import Pg from '../lib/Pg'
import pgconnection from '../lib/pgconnection'

import couchdocs from './mocks/docs.json'

describe('pg', () => {
  const COUCH_URL = 'http://admin:pass@localhost:5984/something'
  const PG_URL = 'postgres://localhost:5432/pg-test'
  const DEFAULT_DATABASE = 'default-database'
  const pg = new Pg(PG_URL, COUCH_URL)

  beforeEach(async () => await pg.drop())
  afterEach(async () => await pg.drop())
  afterAll(async () => await pg.destroy())

  test('invalid constructor', async ()=> {
    expect(() => new Pg()).toThrowError(/Missing Parameter/)
    expect(() => new Pg(PG_URL)).toThrowError(/Missing Parameter/)
    expect.assertions(2)
  })

  test('insert, docs, delete, count', async () => {
    await migrate(PG_URL)

    await pg.insert(couchdocs)
    const docs = await pg.docs()
    expect(docs.length).toBe(2)
    expect(docs).toMatchSnapshot()
    await pg.delete([docs[0].doc._id])
    expect(await pg.count()).toBe(1)
    await pg.delete([])
    expect(await pg.count()).toBe(1)
  })

  describe('sequences', () => {
    const PG2_URL = 'postgres://localhost:5432/pg-test2'
    const pg2 = new Pg(PG2_URL, 'http://localhost:5984/secondpgdb')

    beforeEach(async () => await pg2.drop())
    afterAll(async () => await pg2.destroy())

    test('get and update sequences', async () => {
      await migrate(PG_URL)
      await migrate(PG2_URL)

      const seq = await pg.seq()
      expect(seq).toBe(0)
      await pg.updateSeq(3)
      expect(await pg.seq()).toBe('3')

      const seq2 = await pg2.seq()
      expect(seq2).toBe(0)
      await pg2.updateSeq(13)
      expect(await pg2.seq()).toBe('13')
      expect(await pg.seq()).toBe('3')
    })
  })

  describe('sequences for different scenarios', () => {
    test('existing system with default database', async () => {
      await migrate(PG_URL)
      // Zero sequences
      expect((await pg.sequences()).length).toBe(0)
      // Setup an existing sequence with default-database
      await pg.db('couchdb_progress').insert({seq: '44', source: DEFAULT_DATABASE})

      // Requesting sequence for a given url returns the detault sequence
      // and that sequence source gets updated with this source
      expect(await pg.seq()).toBe('44')
      let sequences = await pg.sequences()
      expect(sequences.length).toBe(1)
      expect(sequences[0].source).toBe(pg.source)
      expect(sequences[0].seq).toBe('44')

      // Requesting a sequence for a second url, creates a new sequence
      // record with seq:0 and source: second-url
      const pg2 = new Pg(PG_URL, 'http://admin:pass@localhost:5984/secondpgdb')
      expect(await pg2.seq()).toBe(0)
      expect(await pg.seq()).toBe('44')
      sequences = await pg.sequences()
      expect(sequences.length).toBe(2)
      expect(sequences[0].source).toBe(pg.source)
      expect(pg.source).toBe('localhost:5984/something')
      expect(sequences[0].seq).toBe('44')
      expect(sequences[1].source).toBe(pg2.source)
      expect(pg2.source).toBe('localhost:5984/secondpgdb')
      expect(sequences[1].seq).toBe('0')
    })

    test('migration to an existing system', async() => {
      await migrate(PG_URL)
      let sequences = await pg.sequences()
      expect(sequences.length).toBe(0)

      await pg.db('couchdb_progress').insert({seq: '44', source: DEFAULT_DATABASE})

      await migrate(PG_URL)
      sequences = await pg.sequences()
      expect(sequences.length).toBe(1)
      expect(sequences[0].source).toBe(DEFAULT_DATABASE)
      expect(sequences[0].seq).toBe('44')
    })
  })
})
