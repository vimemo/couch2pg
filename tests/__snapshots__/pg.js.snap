// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pg insert, docs, sortedDocs, delete, count 1`] = `
Array [
  anonymous {
    "doc": Object {
      "_id": "015ebd5d-5452-4f12-ba30-d41fe4ded03c",
      "_rev": "2-cae134713cab00d1c0368597ca4dd728",
      "alternate_phone": "",
      "date_of_birth": "1994-11-20",
      "date_of_birth_method": "approx",
      "debug_note": "",
      "name": "Amanda Allen",
      "notes": "",
      "parent": Object {
        "_id": "c3ced356-f845-4222-b107-e2568feff2c7",
        "parent": Object {
          "_id": "a0689548-8ade-4b55-b691-074ee548a0d4",
          "parent": Object {
            "_id": "18bf0bfe-d38a-4961-9134-add156b1c3a5",
          },
        },
      },
      "patient_id": "56650",
      "phone": "+14155556655",
      "pregnant_at_registration": "false",
      "reported_date": 1511153396479,
      "role": "patient",
      "sex": "female",
      "transitions": Object {
        "generate_patient_id_on_people": Object {
          "last_rev": 2,
          "ok": true,
          "seq": 1327,
        },
      },
      "type": "person",
    },
  },
  anonymous {
    "doc": Object {
      "_id": "0186101b-be93-4be3-97bd-4ace7f4b94dd",
      "_rev": "2-c50fafe8c58f7afa01666a573b5a18d8",
      "imported_date": "2018-01-21T17:03:55.864Z",
      "name": "Dyah Wahyuningsih",
      "parent": Object {
        "_id": "f02aba23-3219-42de-9add-497baca490f1",
        "parent": Object {
          "_id": "35bf7334-7d45-4763-b67f-f7d98abf8dbb",
          "parent": Object {
            "_id": "2e724ea8-cedb-4d31-95dc-632710c4f571",
          },
        },
      },
      "patient_id": "02732",
      "phone": "+628568408080",
      "reported_date": 1516554238437,
      "transitions": Object {
        "generate_patient_id_on_people": Object {
          "last_rev": 2,
          "ok": true,
          "seq": 2285,
        },
      },
      "type": "person",
    },
  },
]
`;
